[Utils] adds humanize_number method to chat formatting

Drapersniper opened the pull request:
### Type

- [ ] Bugfix
- [X] Enhancement
- [X] New feature

### Description of the changes
This PR adds a new method to chat formatting that allows formatting int and float based on bot locale.

e.g. 1456425612 will now be shows as 1,456,425,612 on en_US

I've applied this method to all mentions of currency I came across (Audio, Bank, Economy, Trivia)

Commit:
[Audio] Add a local cache to avoid making repeated API calls to YouTube API

Commit:
[Audio] Add a local cache to avoid making repeated API calls to YouTube API

Commit:
Merge remote-tracking branch 'origin/V3/develop' into V3/develop

Commit:
[Context] Adds react_quietly method to context

This allows developers to emojis to command messages
This is a modified version of tick()

It accepts True/False for Tick/Cross emoji and in addition to that any other emoji the bot can see

Commit:
Think I may be missing some brain cells

Commit:
Bring back in line with Red

Commit:
[Currency] Add locale aware humanize_int
Applies humanize_int to all (hopefully) mentions of currencies
Adds humanize_int method to chat_formatting so that cog creators can use it as well

Drapersniper committed:
Update setup.cfg

Drapersniper committed:
Update chat_formatting.py

Commit:
[Currency] Add locale aware humanize_int
Applies humanize_int to all (hopefully) mentions of currencies
Adds humanize_int method to chat_formatting so that cog creators can use it as well

Commit:
[Currency] Add locale aware humanize_int
Applies humanize_int to all (hopefully) mentions of currencies
Adds humanize_int method to chat_formatting so that cog creators can use it as well

Commit:
Update Black formatting + 1 extra currency mention in bank

Commit:
When you confuse cog bank with core bank .... (Black formatting)

Commit:
Update one number conversion that already existed in errors
It previously formatted it as "{int:,}"

Drapersniper committed:
So had to add a way to convert xx-XX locales to xx_XX as babel only works with the latter
Tested this and its working with
xx-XX
xx_XX
xx-XX.encoding

Which should work with all locales supported by the bot

Tobotimus commented on redbot/core/utils/chat_formatting.py:
```diff
@@ -429,3 +430,44 @@ def humanize_timedelta(
+    str
+        locale aware formated number.
+    """
+    return format_decimal(val, locale=_convertlocale(get_locale()))
```
There's a chance the locale being used isn't actually known to Babel, in which case it will throw a `babel.core.UnknownLocaleError` - when this happens, simply default to `en_US` :smiley:

Drapersniper replied:
Just for wider visibility,

I changed the code up a little bit.

But basically it uses the parses provided by babel (no need to reinvent the wheel) 
It first tries to parse the red locale, if it fails with `ValueError`, then it tries to get the closest match, if  it  fails (`ValueError` or `TypeError` or `babel.core.UnknownLocaleError`) or can't get a close match then it defaults to  `en_US`

Red's locale and Babel supported locale is then cached and only updated when Red's locale is changed
This avoid calling the babel parser and Try/Except blocks every single time we call `humanize_int`

Tried with about 100 different locale formats and not once it was defaulted to `en_US` :D 

Note to self: Stop trying to reinvent the wheel

Tobotimus reviewed (changes requested):
This looks great! Just one change being requested for now :smile:

I'll also add that I think `humanize_decimal` might be a better name for the util, considering it works for any decimal number (int or float)

Drapersniper committed:
Removed `_convertlocale` instead of trying to reinvent the wheel

Created `_get_babel_locale` which will now only be run on bot start up and locale change

`_get_babel_locale` will try to parse Red's locale to a value supported by `babel`
If it fails (`ValueError`) then it will try to get the closes match locale
If this fails (`ValueError`, `TypeError`, `babel.core.UnknownLocaleError`) then then the locale defaults to `en-US`
The babel locale is cached for faster calls going forward until Red's locale changes

Drapersniper committed:
Made `get_babel_locale(locale: Optional[str] = None`  which converts a string locale to `babel.core.Locale` object
Is the locale is left none it uses Red's locale

Moved `get_babel_locale` and `_get_babel_locale` to to i18n

`humanize_int` now always uses the bots locale

Drapersniper commented:
Thanks @Tobotimus for the suggestions via DM made this a lot more useful!

I may or many not have only seen your mention of `humanize_decimal` its true, but decimal may confuse less familiar devs ? maybe `humanize_number` ?  (however that kinda implies that string numbers also work, im not agaisn't trying to convert a number to int/float if i fails raise it back to user) i'll let you confirm the desired aproach here.

skeith commented:
I'd say `humanize_digit` since it's actually called digit grouping separator.

Tobotimus commented:
I think `humanize_number` is best, I guess it's implied that the number will be formatted as a decimal :)

There shouldn't be any need to handle string inputs, simply state the argument type in the docs and users will figure it out.

`humanize_digit` is confusing to me, because a digit refers to a single character in a number

Drapersniper committed:
Updated method name to `humanize_number` as it works with both float and ints

Drapersniper committed:
Refaction calls to `humanize_int` to `humanize_number`

Drapersniper committed:
3rd time lucky? (black formatting)

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Add towncrier entries

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Add towncrier entries

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Tobotimus commented on changelog.d/2836.feature.rst:
```diff
@@ -0,0 +1 @@
+New :meth:`humanize_number` in :module:`redbot.core.utils.chat_formatting` function to convert numbers into text which respect locale.
```
Should be ``:func:`humanize_number` ``, `:meth:` is for methods.
```suggestion
New :func:`humanize_number` in :module:`redbot.core.utils.chat_formatting` function to convert numbers into text which respect locale.
```

Tobotimus commented on changelog.d/2836.misc.rst:
```diff
@@ -0,0 +1 @@
+New :meth:`humanize_number` is used throughout the bot.
```
```suggestion
New :func:`humanize_number` is used throughout the bot.
```

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -152,7 +155,7 @@ def check(m):
 
         reason = "{}({}) deleted {} messages containing '{}' in channel {}.".format(
-            author.name, author.id, len(to_delete), text, channel.id
+            author.name, author.id, humanize_number(len(to_delete)), text, channel.id
```
This one should probably use `babel.format_decimal` directly with en_US as the locale, since the message will always be in English, we probably shouldn't put a number formatted for a different locale in there. Especially when we have context-specific locale :)

Drapersniper replied:
I've added the `override_locale` arg to `humanize_number` this is to facilitate the overriding of the current bots locale while still being able to use `humanize_number` (So devs don't need to import babel and figure out it and do all the error handling we done here)

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -208,7 +211,14 @@ def check(m):
+            "".format(
+                author.name,
+                author.id,
+                humanize_number(len(to_delete)),
```
Same as previous comment

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -240,7 +250,7 @@ def check(m):
 
         reason = "{}({}) deleted {} messages in channel {}.".format(
-            author.name, author.id, len(to_delete), channel.name
+            author.name, author.id, humanize_number(len(to_delete)), channel.name
```
Same here

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -277,7 +287,7 @@ def check(m):
 
         reason = "{}({}) deleted {} messages in channel {}.".format(
-            author.name, author.id, len(to_delete), channel.name
+            author.name, author.id, humanize_number(len(to_delete)), channel.name
```
Same here

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -319,7 +329,7 @@ def check(m):
         to_delete.append(ctx.message)
         reason = "{}({}) deleted {} messages in channel {}.".format(
-            author.name, author.id, len(to_delete), channel.name
+            author.name, author.id, humanize_number(len(to_delete)), channel.name
```
Same here

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -420,7 +430,7 @@ def check(m):
             "{}({}) deleted {} "
             " command messages in channel {}."
-            "".format(author.name, author.id, len(to_delete), channel.name)
+            "".format(author.name, author.id, humanize_number(len(to_delete)), channel.name)
```
Same here

Tobotimus commented on redbot/cogs/cleanup/cleanup.py:
```diff
@@ -500,7 +510,7 @@ def check(m):
             "{}({}) deleted {} messages "
             "sent by the bot in {}."
-            "".format(author.name, author.id, len(to_delete), channel_name)
+            "".format(author.name, author.id, humanize_number(len(to_delete)), channel_name)
```
Same here

Tobotimus commented on redbot/cogs/economy/economy.py:
```diff
@@ -284,9 +289,9 @@ def __init__(self, bot: Red):
-                        pos=pos,
+                        amount=humanize_number(await self.config.PAYDAY_CREDITS()),
+                        new_balance=humanize_number(await bank.get_balance(author)),
+                        pos=humanize_number(pos) if pos else pos,
```
When can pos be zero? If there is a legitimate case for this, please add a comment explaining it :)

Drapersniper replied:
in `get_leaderboard_position` there is a `return None` so there is a chance of it being None rather than 0 , it seems to use `discord.utils.find` which can return None... Doubt it could ever happen from a quick glance at the code, but since it's in the source function just wanted to handle it.

Tobotimus replied:
Ah no problem :+1:

Tobotimus commented on redbot/cogs/economy/economy.py:
```diff
@@ -331,9 +338,9 @@ def __init__(self, bot: Red):
-                        pos=pos,
+                        amount=humanize_number(credit_amount),
+                        new_balance=humanize_number(await bank.get_balance(author)),
+                        pos=humanize_number(pos) if pos else pos,
```
Same as previous comment

Tobotimus commented on redbot/cogs/economy/economy.py:
```diff
@@ -384,14 +391,14 @@ def __init__(self, bot: Red):
 
             if acc[0] != author.id:
-                temp_msg += f"{f'{pos}.': <{pound_len+2}} {balance: <{bal_len + 5}} {name}\n"
+                temp_msg += f"{f'{humanize_number(pos)}.': <{pound_len+2}} {balance: <{bal_len + 5}} {name}\n"
```
Line is a bit too long, could you split it over two lines? Black doesn't split up strings to get under the max line length

Tobotimus commented on redbot/core/modlog.py:
```diff
@@ -149,7 +150,9 @@ def __init__(
         title = "{}".format(
-            _("Case #{} | {} {}").format(self.case_number, casetype.case_str, casetype.image)
+            _("Case #{} | {} {}").format(
+                humanize_number(self.case_number), casetype.case_str, casetype.image
```
I think in this circumstance we shouldn't use `humanize_number`. We use case numbers as input to other commands, this makes it slightly more annoying for users to copy+paste the case number

Tobotimus reviewed (changes requested):
Just a few small changes for now :)

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Update PR to reflect Toby's review

Signed-off-by: Draper <guyreis96@gmail.com>

Drapersniper reviewed (commented):
I've also added the Override to en_US on a few errors and and log messages that you didn't mention on the review, let me know if you'd prefer reverting it.

Drapersniper commented on redbot/core/bank.py:
```diff
@@ -242,12 +242,19 @@ def _invalid_amount(amount: int) -> bool:
         raise TypeError("Withdrawal amount must be of type int, not {}.".format(type(amount)))
     if _invalid_amount(amount):
-        raise ValueError("Invalid withdrawal amount {} < 0".format(humanize_number(amount)))
+        raise ValueError(
```
Wasn't too sure if this gets translated to users locale down the pipe never used the bot with a non en locale.
If it doesn't it make sense to override these to en_US, but if its translated later then this should be removed

Tobotimus replied:
That's all good, you made the right call :+1: Even if these messages are sent to the user, they're not translated, so using `humanize_number` with a custom locale wouldn't make sense.

Drapersniper commented on redbot/core/bank.py:
```diff
@@ -242,12 +242,19 @@ def _invalid_amount(amount: int) -> bool:
     if amount > bal:
         raise ValueError(
-            "Insufficient funds {} > {}".format(humanize_number(amount), humanize_number(bal))
+            "Insufficient funds {} > {}".format(
```
Same as above

Drapersniper commented on redbot/core/bank.py:
```diff
@@ -279,7 +286,11 @@ def _invalid_amount(amount: int) -> bool:
         raise TypeError("Deposit amount must be of type int, not {}.".format(type(amount)))
     if _invalid_amount(amount):
-        raise ValueError("Invalid deposit amount {} <= 0".format(humanize_number(amount)))
+        raise ValueError(
```
Same as above

Drapersniper commented on redbot/core/bank.py:
```diff
@@ -313,7 +324,11 @@ def _invalid_amount(amount: int) -> bool:
         raise TypeError("Transfer amount must be of type int, not {}.".format(type(amount)))
     if _invalid_amount(amount):
-        raise ValueError("Invalid transfer amount {} <= 0".format(humanize_number(amount)))
+        raise ValueError(
```
Same as above

Drapersniper committed:
Add missing stop

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Tobotimus commented on redbot/core/i18n.py:
```diff
@@ -160,6 +171,32 @@ def _add_translation(self, untranslated, translated):
+    return babel_locale
+
+
+def get_babel_locale(locale: Optional[str] = None) -> babel.core.Locale:
```
Could you add a docstring for this function? And make sure it's in the i18n docs :)

Drapersniper replied:
![image](https://user-images.githubusercontent.com/27962761/63760362-134e9c00-c8b7-11e9-8258-58673caf481b.png)
Hows this?

Drapersniper committed:
Merge branch 'V3/develop' into humanize_int

Drapersniper committed:
Add docstrings to `get_babel_locale`

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Fix some typos

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Tobotimus committed:
Merge branch 'V3/develop' into pull/2836

# Conflicts:
#	redbot/cogs/economy/economy.py
